### PR TITLE
Fix and improve example : with-componentdidcatch

### PR DESCRIPTION
**Pull request by lucleray**
In this PR, I wrote 2 commits :

- https://github.com/zeit/next.js/commit/b6d2a727b9e9bd60295a9322eeb61cf22e09b051 Fix the example, by throwing an error only on the client-side (thanks @timneutkens for your help on that 😁)

- https://github.com/zeit/next.js/commit/b08863c7a98c8a869293b882008700c9da24e732 Improve the example. I transposed the [error boundaries demo](https://codepen.io/gaearon/pen/wqvxGa?editors=0010) in [react docs](https://reactjs.org/docs/error-boundaries.html#live-demo) to Next.js.

### CODE REVIEW

File: examples/with-componentdidcatch/package.json [hunk 1]
```diff
@@ -7,7 +7,7 @@
     "start": "next start"
   },
   "dependencies": {
-    "next": "6.0.0-canary.6",
+    "next": "6.0.3",
```
Comment (timneutkens) [inline examples/with-componentdidcatch/package.json]: Awesome, but let's use `latest` for this now 🙏

File: examples/with-componentdidcatch/pages/_app.js [hunk 2]
```diff
@@ -1,9 +1,48 @@
...
+  state = { error: null, errorInfo: null }
+
   componentDidCatch (error, errorInfo) {
-    console.log('CUSTOM ERROR HANDLING', error)
-    // This is needed to render errors correctly in development / production
-    super.componentDidCatch(error, errorInfo)
+    this.setState({ error, errorInfo })
+
+    // If you wish to render error with the usual error page, you can use :
+    // super.componentDidCatch(error, errorInfo)
```
Comment (timneutkens) [inline examples/with-componentdidcatch/pages/_app.js]: This is very important in development mode, where you want the debug component to render with all the information and source maps applied on the stack traces

Comment (timneutkens) [reply examples/with-componentdidcatch/pages/_app.js]: In production this will make sure _error.js is rendered.

Comment (timneutkens) [reply examples/with-componentdidcatch/pages/_app.js]: Which is the reason I only did a `console.log` in the example

Comment (lucleray) [reply examples/with-componentdidcatch/pages/_app.js]: That makes sense. It's probably better to edit _error.js than handling the error here.

Comment (timneutkens) [reply examples/with-componentdidcatch/pages/_app.js]: Yep!

Comment (lucleray) [reply examples/with-componentdidcatch/pages/_app.js]: I'm reverting _app.js back to its previous state 😙

File: examples/with-componentdidcatch/pages/index.js [hunk 3]
```diff
@@ -1,4 +1,30 @@
-export default () => {
-  // Render time error
-  throw new Error('Test')
+import React, { Component } from 'react'
+
+class BuggyCounter extends Component {
```
Comment (timneutkens) [inline examples/with-componentdidcatch/pages/index.js]: Awesome class name :joy:

**Review by timneutkens — APPROVED** [review]

File: examples/with-componentdidcatch/pages/index.js (hunk 3 above)
Comment (lucleray) [reply examples/with-componentdidcatch/pages/index.js]: 😂

(credit goes to [this demo](https://codepen.io/gaearon/pen/wqvxGa?editors=0010) from the react docs)

**Pull request merged**